src/doc/rgw/user-accounts.md

```diff
@@ -0,0 +1,58 @@
+# iam accounts
+
+add a user account primitive for use with IAM REST APIs. this enables end-users to use tools like `aws iam create-user` for self-service management of the users, roles, policies, and access keys in their account without requiring global admin privileges for radosgw-admin commands or radosgw /admin REST APIs
+
+add new IAM APIs for User and AccessKey operations. some Role/RolePolicy/UserPolicy operations are already supported, but require special admin capabilities. once a User or Role are added to an account, that account's root user will be granted full access to them
+
+existing users and roles must be added to an account manually, either by a radosgw-admin command or /admin API, before they can be managed by an account root user
+
+
+## namespacing
+
+rgw supports namespace isolation of users and buckets using the 'tenant' name, such that two tenants A and B can both have buckets named 'example-bucket', or users named 'example-user', that are completely indepedent of one another
+
+there exists no 'tenant' metadata that we can attach policy to, index users or buckets on, or accumulate usage stats for reporting or quota enforcement. features like this will instead be attached to the proposed account metadata
+
+to preserve this namespacing functionality, each account can optionally be associated with a tenant name. all users and roles in that account must have that same tenant name
```

Comment:
Can multiple accounts have the same tenant?  Or is tenant a 1-to-1 with an account?

Reply:
for ease of understanding, we should clarify the implications of accounts sharing a tenant and maybe why someone would do it--iirc, @mmgaggle has described the ability to maintain a hierarchy of accounts in aws, and I don't think we're bringing all that detail in (e.g.)

Reply:
> we should clarify the implications of accounts sharing a tenant and maybe why someone would do it

you'd use multiple tenants if you want namespace isolation, and multiple accounts if you want isolation of user/role/policy management

if you want a global dns namespace like aws, put all accounts in the global empty-tenant namespace

> iirc, @mmgaggle has described the ability to maintain a hierarchy of accounts in aws, and I don't think we're bringing all that detail in (e.g.)

interested to hear more about that requirement and how it relates to tenant namespacing. naively, it sounds like we'd just want an account to have an optional 'parent' account that it can inherit some policy from

Reply:
> > iirc, @mmgaggle has described the ability to maintain a hierarchy of accounts in aws, and I don't think we're bringing all that detail in (e.g.)
> 
> interested to hear more about that requirement and how it relates to tenant namespacing. naively, it sounds like we'd just want an account to have an optional 'parent' account that it can inherit some policy from

in aws, it's the AWS Organizations API that's responsible for the grouping of related accounts and the management of policy like service control policy:

https://docs.aws.amazon.com/organizations/latest/APIReference/Welcome.html
https://docs.aws.amazon.com/organizations/latest/userguide/orgs_introduction.html
https://docs.aws.amazon.com/organizations/latest/userguide/orgs_manage_policies_scps.html

maybe we could find some subset of that Organizations API that would accomplish our goals here, instead of adding special custom behaviors to the accounts themselves?

Reply:
I agree @cbodley; using a subset of the OUs API could help us introduce a similar concept as to what AWS does with Control Tower, specifically with what they call preventive controls, which looks like it's enforcing SCPs at the OU level , perhaps within organizations, we could introduce templates for different policy/configurations and rules for enforcing them.